### ISSUE TITLE
How can I obtain loss values for testing?

### ISSUE THREAD

**Issue by DilaraAlbayrak** [junyanz/pytorch-CycleGAN-and-pix2pix#1009]

**Comment by hyecheol123** [comment]
I also have the same question!!

**Comment by hyecheol123** [comment]
#229 According to the previous issue (approximately 2 years ago), it says it is not supported.
Is this functionality still not supported?

**Comment by junyanz** [comment]
It is not supported. We might add evaluation test time metrics (FID/meanAP) in the future project release. But we don't plan to add this functionality now.

**Comment by hyecheol123** [comment]
If I want to make those by myself, will it be okay just to copy and paste the relevant code sections from the training.py?

**Comment by junyanz** [comment]
Sure. You can modify the `test` or `forward` [function](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/base_model.py#L98).

**Comment by hyecheol123** [comment]
How can I get the loss for each model structure?
Feels like I need to define/modify some functions on pix2pix_model.py to do so, right?

As we do not load discriminator, we can only calculate L1 loss, right? I want to verify whether my understanding is correct or not.

What I will do follows:
 - define criterionL1() for testing (only when we have paired images for testing)
 - Get all L1 losses for the testing set, and calculate mean, sd, ...
 - Print the statistics to console

**Comment by junyanz** [comment]
Yes. You need to modify the `pix2pix_model.py` as each model has different losses. You can calculate the mean in the `test.py` after the main for loop. If you want to load D, you can modify this [line](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/models/pix2pix_model.py#L54).

**Comment by hyecheol123** [comment]
@junyanz I successfully calculated the L1 loss, and I am now trying to calculate discriminator loss after loads D.

Do you want me to create a PR after I finish that? If you need the functionality, I am happy to do so, but otherwise, as I understand creating unhelpful PR is a waste of time and manpower, I will not submit one. How do you think?

**Comment by junyanz** [comment]
Great. A PR might involve more work, as we want an abstraction that works for all the models and losses. If you want to give a try, we can have a look at PR. But no pressure. We will revisit it by ourselves later when we have some cycle.

**Comment by hyecheol123** [comment]
@junyanz I try to calculate discriminator loss, but when I enable discriminator loss calculation, it makes weird testing results image (Get image with random noise).

Can you help me with how to handle this problem??

```
        # Calculate L1 loss while testing if needed
        if not(self.isTrain):
            if self.opt.test_loss == 1:
                self.loss_G_L1 = self.criterionL1(self.fake_B, self.real_B) * self.opt.lambda_L1
                # Calculate Discriminator Loss & GAN Loss
                fake_AB = torch.cat((self.real_A, self.fake_B), 1)
                pred_fake = self.netD(fake_AB.detach())
                self.loss_D_fake = self.criterionGAN(pred_fake, False)
                real_AB = torch.cat((self.real_A, self.real_B), 1)
                pred_real = self.netD(real_AB)
                self.loss_D_real = self.criterionGAN(pred_real, True)
                pred_fake = self.netD(fake_AB)
                self.loss_G_GAN = self.criterionGAN(pred_fake, True)
```

**Comment by hyecheol123** [comment]
I checked the backward part and select the code that I needed for calculating the loss.

**Comment by hyecheol123** [comment]
I found the reason why.
I did not specify the network name ['G', 'D'] when I initialize the network. I only use ['G'], and I successfully fixed it.

Once I finish updating my project's code, then I will work on modifying the model based on this code as my project code has been modified a lot (removed unnecessary part from the original code).

It may take some time to give you a new PR regarding this issue, but I will do it as soon as I can.

**Comment by zhouyingji** [comment]
> I found the reason why.
> I did not specify the network name ['G', 'D'] when I initialize the network. I only use ['G'], and I successfully fixed it.
> Once I finish updating my project's code, then I will work on modifying the model based on this code as my project code has been modified a lot (removed unnecessary part from the original code).
> It may take some time to give you a new PR regarding this issue, but I will do it as soon as I can.

If your code finish? Where I can find your L1 loss calculate code? Thank you!

**Comment by hyecheol123** [comment]
@zhouyingji
As I forked and modified a lot of irrelevant parts of the original code (this repository) for my own project, I decided to fork again this repo and only fix the relevant part of the code to get the loss values on testing time before I submit a PR.

Sorry for being late due to my spring semester's final and the start of the summer session. I promise it won't take much time as I already finished that the code is working properly (only for pix2pix, though).

**Comment by zhouyingji** [comment]
@hyecheol123  
Thanks for your reply and plz tell me if your PR submit. My own project based on pix2pix and I hope calculate L1 loss when testing. ActuIly I don't care about discriminator loss when I translate the image using latest generator.
